optimizely/config_manager.py

```diff
@@ -0,0 +1,233 @@
+
+    @staticmethod
+    def get_datafile_url(sdk_key, url, url_template):
+        """ Helper method to determine URL from where to fetch the datafile.
+
+        Args:
+          sdk_key: Key uniquely identifying the datafile.
+          url: String representing URL from which to fetch the datafile.
+          url_template: String representing template which is filled in with
+                        SDK key to determine URL from which to fetch the datafile.
+
+        Returns:
+          String representing URL to fetch datafile from.
+
+        Raises:
+          optimizely.exceptions.InvalidInputException if:
+          - One of sdk_key or url is not provided.
+          - url_template is invalid.
+        """
+        # Ensure that either is provided by the user.
+        if sdk_key is None and url is None:
+            raise optimizely_exceptions.InvalidInputException('Must provide at least one of sdk_key or url.')
+
+        # Return URL if one is provided or use template and SDK key to get it.
+        if url is None:
+            try:
+                return url_template.format(sdk_key=sdk_key)
+            except (AttributeError, KeyError):
+                raise optimizely_exceptions.InvalidInputException(
+                    'Invalid url_template {} provided.'.format(url_template))
+
+        return url
+
+    def set_update_interval(self, update_interval):
+        """ Helper method to set frequency at which datafile has to be polled and ProjectConfig updated. """
+        self.update_interval = update_interval or enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+        # If polling interval is less than minimum allowed interval then set it to default update interval.
+        if self.update_interval < enums.ConfigManager.MIN_UPDATE_INTERVAL:
+            self.logger.debug('Invalid update_interval {} provided. Defaulting to {}'.format(
+                update_interval,
+                enums.ConfigManager.DEFAULT_UPDATE_INTERVAL)
+            )
+            self.update_interval = enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    def set_last_modified(self, response_headers):
+        """ Looks up and sets last modified time based on Last-Modified header in the response.
+
+         Args:
+           response_headers: requests.Response.headers
+         """
+        self.last_modified = response_headers.get(enums.HTTPHeaders.LAST_MODIFIED)
+
+    def set_config(self, datafile):
+        """ Looks up and sets datafile and config based on response body.
+
+         Args:
+           datafile: JSON string representing the Optimizely project.
+         """
+        # TODO(ali): Add validation here to make sure that we do not update datafile and config if not a valid datafile.
+        self._datafile = datafile
+        # TODO(ali): Add notification listener.
+        self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+        self.logger.info('Received new datafile and updated config.')
```

Comment:
I think this could just be a `debug`. Info is kinda noisy if we expect datafile updates to be happening recurringly

Reply:
Yes. Good catch.